core/server/api/authentication.js

```diff
@@ -244,39 +258,105 @@ authentication = {
+                });
+        }
+
+        function extractTokenParts (options) {
+            var data = options.data.passwordreset[0];
+
+            tokenParts = globalUtils.tokens.resetToken.extract({
+                token: data.token
+            });
+
+            if (!tokenParts) {
+                return Promise.reject(new errors.UnauthorizedError({
+                    message: 'Invalid token'
+                }));
+            }
 
-        function validateRequest(resetRequest) {
-            return utils.checkObject(resetRequest, 'passwordreset');
+            return Promise.resolve();
         }
 
-        function doReset(resetRequest) {
-            var settingsQuery = {context: {internal: true}, key: 'dbHash'},
-                data = resetRequest.passwordreset[0],
+        // @TODO: use brute force middleware (see https://github.com/TryGhost/Ghost/pull/7579)
+        function protectBruteForce() {
+            if (tokenSecurity[tokenParts.email + '+' + tokenParts.expires] &&
+                tokenSecurity[tokenParts.email + '+' + tokenParts.expires].count >= 10) {
+                return Promise.reject(new errors.NoPermissionError({
+                    message: i18n.t('errors.models.user.tokenLocked')
+                }));
+            }
+
+            return Promise.resolve();
+        }
+
+        function doReset(options) {
+            var data = options.data.passwordreset[0],
                 resetToken = data.token,
+                oldPassword = data.oldPassword,
                 newPassword = data.newPassword,
-                ne2Password = data.ne2Password;
-
-            return settings.read(settingsQuery).then(function then(response) {
-                return models.User.resetPassword({
-                    token: resetToken,
-                    newPassword: newPassword,
-                    ne2Password: ne2Password,
-                    dbHash: response.settings[0].value
+                user;
+
+            return settings.read(_.merge({key: 'dbHash'}, options))
+                .then(function fetchedSettings(response) {
+                    dbHash = response.settings[0].value;
+
+                    return models.User.getByEmail(tokenParts.email, options);
+                })
+                .then(function fetchedUser(_user) {
+                    user = _user;
+
+                    if (!user) {
+                        throw new errors.NotFoundError({message: i18n.t('errors.api.users.userNotFound')});
+                    }
+
+                    tokenIsCorrect = globalUtils.tokens.resetToken.compare({
+                        token: resetToken,
+                        dbHash: dbHash,
+                        password: user.get('password')
+                    });
+
+                    if (!tokenIsCorrect) {
+                        return Promise.reject(new errors.BadRequestError({
+                            message: i18n.t('errors.models.user.invalidTokenStructure')
```

Comment:
see key change in commit
https://github.com/TryGhost/Ghost/pull/7622/commits/5f122832ee38d1cf151439ceb79c2d4d1910a462